daprovider/server/provider_server.go

```diff
@@ -0,0 +1,193 @@
+	"github.com/offchainlabs/nitro/daprovider/daclient"
+)
+
+type Server struct {
+	reader    daprovider.Reader
+	writer    daprovider.Writer
+	validator daprovider.Validator
+}
+
+type ServerConfig struct {
+	Addr               string                              `koanf:"addr"`
+	Port               uint64                              `koanf:"port"`
+	JWTSecret          string                              `koanf:"jwtsecret"`
+	EnableDAWriter     bool                                `koanf:"enable-da-writer"`
+	ServerTimeouts     genericconf.HTTPServerTimeoutConfig `koanf:"server-timeouts"`
+	RPCServerBodyLimit int                                 `koanf:"rpc-server-body-limit"`
+}
+
+var DefaultServerConfig = ServerConfig{
+	Addr:               "localhost",
+	Port:               9880,
+	JWTSecret:          "",
+	EnableDAWriter:     false,
+	ServerTimeouts:     genericconf.HTTPServerTimeoutConfigDefault,
+	RPCServerBodyLimit: genericconf.HTTPServerBodyLimitDefault,
+}
+
+func ServerConfigAddOptions(prefix string, f *flag.FlagSet) {
+	f.String(prefix+".addr", DefaultServerConfig.Addr, "JSON rpc server listening interface")
+	f.Uint64(prefix+".port", DefaultServerConfig.Port, "JSON rpc server listening port")
+	f.String(prefix+".jwtsecret", DefaultServerConfig.JWTSecret, "path to file with jwtsecret for validation")
+	f.Bool(prefix+".enable-da-writer", DefaultServerConfig.EnableDAWriter, "implies if the das server supports daprovider's writer interface")
+	f.Int(prefix+".rpc-server-body-limit", DefaultServerConfig.RPCServerBodyLimit, "HTTP-RPC server maximum request body size in bytes; the default (0) uses geth's 5MB limit")
+	genericconf.HTTPServerTimeoutConfigAddOptions(prefix+".server-timeouts", f)
+}
+
+func fetchJWTSecret(fileName string) ([]byte, error) {
+	if data, err := os.ReadFile(fileName); err == nil {
+		jwtSecret := common.FromHex(strings.TrimSpace(string(data)))
+		if len(jwtSecret) == 32 {
+			log.Info("Loaded JWT secret file", "path", fileName, "crc32", fmt.Sprintf("%#x", crc32.ChecksumIEEE(jwtSecret)))
+			return jwtSecret, nil
+		}
+		log.Error("Invalid JWT secret", "path", fileName, "length", len(jwtSecret))
+		return nil, errors.New("invalid JWT secret")
+	}
+	return nil, errors.New("JWT secret file not found")
+}
+
+// NewServerWithDAPProvider creates a new server with pre-created reader/writer/validator components
+func NewServerWithDAPProvider(ctx context.Context, config *ServerConfig, reader daprovider.Reader, writer daprovider.Writer, validator daprovider.Validator) (*http.Server, error) {
+	listener, err := net.Listen("tcp", fmt.Sprintf("%s:%d", config.Addr, config.Port))
+	if err != nil {
+		return nil, err
+	}
+
+	rpcServer := rpc.NewServer()
+	if config.RPCServerBodyLimit > 0 {
+		rpcServer.SetHTTPBodyLimit(config.RPCServerBodyLimit)
+	}
+
+	server := &Server{
+		reader:    reader,
+		writer:    writer,
+		validator: validator,
+	}
+	if err = rpcServer.RegisterName("daprovider", server); err != nil {
```

Comment:
Since you have 3 separate interfaces (reader/writer/validator) and each has their own rpc functions - it might make sense to have 3 separate provider server objects - one for wrapping each interface - and to have here 3 separate calls to rpcServer.RegisterName - even if all 3 objects register to the same namespace.
Not a must.

Reply:
I haven't done this yet but it should be an easy enough change. Either tomorrow or in a future PR.

Reply:
I'm already dreading the merge conflicts of this PR back to custom-da especially around provider_server so I think I'll defer this change til after this is merged to master, and then master is merged back to custom-da.

Reply:
Fixed on `custom-da` in https://github.com/OffchainLabs/nitro/pull/3237/commits/87498333f05e4e062c94daa65ed0f4d516154cdb